### ISSUE TITLE
Stoploss or trailing stoploss, what happen first in backtest? 

### ISSUE THREAD

**Issue by Fimap9to1** [freqtrade/freqtrade#4105]
Hi, I was testing the strategy live and realized that many stoploss (fixed) are being hit comparing with backtest. 

I use trailing stoploss as a sell signal and it seems like in backtesting this has the priority over the fixed stoploss.

I've read the assumpions on backtesting but found this information. Actually it seems like that, but I'd like it to make sure, thanks!

Below the assumptions I've read about backtesting:

FIXED STOPLOSS
Low happens before high for stoploss, protecting capital first.

TRAILING STOPLOSS
High happens first - adjusting stoploss
Low uses the adjusted stoploss (so sells with large high-low difference are backtested correctly)

So, which one happens first, low or high if I use trailing stoploss?

Thaks for your answer!

**Comment by xmatthias** [comment]
well - you even wrote the answer yourself (but maybe i'm not understanding the question ... as i doubt you'll ask with the answer already included)

TRAILING STOPLOSS
High happens first - adjusting stoploss
Low uses the adjusted stoploss (so sells with large high-low difference are backtested correctly)

**Comment by Fimap9to1** [comment]
Yes I just want to make sure since maybe it's me that's not understanding the sentences.

So do you confirm that in backtesting when we use trailing stoploss, the first data to "happen" is high and not low?

**Comment by xmatthias** [comment]
it is as described in the documentation, if `trailing_stop` is enabled (`trailing_stop=True`), then High happens before Low.

Depending on the values you're using (large distance, small distance) this can work for or against you.

Based on your question, i assume you're using a pretty tight distance.
This will lead to a lot of adjust + adjustedstop situations within one candle - which will seem backtesting work great ... but unfortunately, price patterns within a candle are not that simple, so the stop may trigger before the candle's high is reached.

In general, i'd recommend to look at the general candle length (high / low distance) and have the trailing stop larger than this.

**Comment by Fimap9to1** [comment]
This was very helpfull thank you so much!

**Comment by Fimap9to1** [comment]
It'd be great if trailing wouldn't make high happen first, but leave low happen first and then, if stoploss is not hit, activate trailing from high. This would make backtesting much more reliable.

**Comment by xmatthias** [comment]
unfortunately that's not the case.

assume you use a stoploss of 10%, trailing stop offset of 3%, and a trailing stop of 1%. (so stoploss is moved from 10% based on the open price to a trailing 1% stoploss after a profit of 3% is reached.

---

we're in a "Non-trailing" scenario just yet (profit never reached 3% so far)
* Low is at -2%, high is at 3.5%

Now current behavior is that the stoploss will be adjusted to 3.5% - 1% = 2.5%, and will trigger.

Your propose behavior is to only adjust the price, but don't trigger stoploss until the next candle.
Now the next candle stays between 2.7% and 3.5% (or even raises more) ...
You're still in the trade, and backtesting might even show you a profit of 10% or more assuming the asset raises further ... 

but in reality, you just take a different guess for backtesting.
if the trade had high first, then went from 3.5% to -5% before raising back to 3% , it would trigger in live - so instead of showing a profit of 2.5%, it'll show you a profit of 10% ... this didn't protect capital - but actually made accuracy wrong.

**Comment by Fimap9to1** [comment]
That's not exactly what I meant. For example, let's assume the offset in not activated (as in my case live) to simplify the matter.

We just have a trailing stoploss of 1% and a stoploss (fixed) of 10%. Which is working fair enough, but with an issue in backtesting.

In backtest it's impossible to understand how many stoplosses (fixed) are hit live, since almost all the sell reasons are coming from trailing stoploss (because high happens first).

So in backtesting right now, the trade opens at 0% and then almost all times the high is higher than the open, consequently the trailing is activeted almost on every trade (it's very rare that we don't have a high higher than the open for this trategy but I think this applies for many).

What I suggest is: since in reality we do have many cases in which the price first goes down with little correction and then after that it goes up and make a profit with trailing (without offset still).

Wouldn't it be better if first of all we understand what correction occured and then look for profit?
Because now, if a use a stoploss (fixed) of 20%, it never gets hit. Because in backtest high happen first, activating almost always a trailing stoploss that live would neve be activeted that many time. Instead, in case of murket dump, I would hit so many stoplosses. And in reality i would be rekt if I used that stoploss.

**Comment by xmatthias** [comment]
the only explanation i have for what you describe is a too tight trailing stop / trailing stop offset.

For the rest - see above as to why this behavior is correct in case of a trailing stoploss.

I understand your concerns - but it's pretty rare to see a stoploss of 20% trigger, while at the same candle, a high is raising the stoploss. This would imply that this single candle did at least a -21% move (first +1%, then to -20%) ... which will not happen often unless you use very long candle sizes (in which case, your stoploss / trailing stop must be just as wide).

**Comment by xmatthias** [comment]
Assume the simplest of all cases (i'm using round numbers for simplicity - actual prices in the calculation are slightly different)
This does not respect fees.
stoploss 5%
trailing_stoploss = True

We open the trade at candle open.

this candle looks:
open: 100
high: 102
low: 96
close: 99

Entering this candle - backtesting will adjust stoploss based on high (so it's 102 - 5 = 97).
It'll then trigger the stoploss as trailing-stop at 97.

Your proposing to check stoploss first against low - and only after this adjust stoploss.
The stoploss will therefore not trigger on this candle, and going out of this candle, we've got a stoploss at 97 - but the trade is still open.

This will allow the trade to continue - most likely raising stoploss higher, and therefore inflating profits.

it is therefore not protecting assets first - but inflate profits reports reports, as it'll keep trades open potentially a lot longer than it should.

**Comment by Fimap9to1** [comment]
Ok I get that. And I agree it to be the best tradeoff. 

Thanks again!

**Comment by xmatthias** [comment]
> So basically backtesting using just trailing without offset as a sell signal + stoploss (fixed) is not possible beacause the results are very far from the live.

this is incorrect. 
Results are quite similar as long as you're not using a very tight stoploss.
Obviously by using a stoploss of 1% or closer you can throw of backtesting - but it's exploiting an assumption of backtesting, so can be expected.

**Comment by Fimap9to1** [comment]
In your experience what is a decently reliable trailing stoploss in a 2h candles chart? Movements are not very large generally, exept some cases. So I don't know what to expect live. 

Yes, I'm going to think about that because in my opnion this is particularly important in faster timeframes.

**Comment by Fimap9to1** [comment]
Thanks, actually profits are good even live (many profits and good % per trade). What ruins the result is the number of stoploss (fixed) hit. Which live make the strategy lose (now was -5% and I was thinking about reduce it). But in backtest is not taken into account due to the "high first rule" so I can't decide what is the bes value.

Do you have any idea on how to make the backtest more reliable, even with a low trailing stoploss?

**Comment by xmatthias** [comment]
backtesting is as good as the data available - so for 2h charts, you get 4 datapoints per 2 hour interval. Making it "more reliable" is impossible, as intra-candle moves are not covered.

You could use a lower timeframe (say 5m) and resample (or increase the lookback of the indicator) ... but it'll limit you to 500 5m candles in live - so depending on the indicators used, this may or may not work.
That'll give you visibility into the 2h candles ... 
Freqtrade does intentionally not do this implicitly, as it would increase the duration of backtesting substantially, increase the memory-pressure by a ton (you can backtest 4 years of 2h data, but try that with 5m data) - and it would face a lower limit (usually 1m is the lowest exchanges give us).

**Comment by Fimap9to1** [comment]
Ok understood thanks. Than the max timeperiod available on 5 min is 20 in 2h ((2*20*60)/5) = 480 5min candles right? Is it safe to stay so close to the limit or maybe is better to keep a bigger margin?

**Comment by xmatthias** [comment]
that's about correct (although precise numbers will vary depending on the exchange).

Your problem might be when there was an exchange downtime - as then you won't get 500 candles, but 500 - (downtime duration) ...

**Comment by Fimap9to1** [comment]
I've tried it in 5 min (simply adjusting time periods as mentioned above) and actually me the same result as the 2h one in the days it was losing.

**Comment by Fimap9to1** [comment]
Ok understood about downtime of exchanges :)

**Comment by xmatthias** [comment]
i find both of your trailing stop values too small.

Your offset / positive difference is 0.5% ... assuming you use stoploss on exchange, you're by default giving a 1% margin below the stoploss - so you will sell anywhere between 0.5% and -0.5% (worst case, obviously).

**Comment by Fimap9to1** [comment]
I've changed it this way:

  # Trailing stop:
    trailing_stop = True
    trailing_stop_positive = 0.0185
    trailing_stop_positive_offset = 0.0285
    trailing_only_offset_is_reached = True

But still profitable (live it has lose 317% in one day, worst day) but in backtest the worst day is -56%). I really don't get how this can happen eveng with 5m candles :/

**Comment by xmatthias** [comment]
as mentioned above, tuning trailing stop to be reliable in backtesting is very difficult.

In general, the shorter the timeframe, the better it'll work - but both offset and positive should be > the average timeframe movement  - which it clearly isn't with values of 2% ... 

In general, i'd encourage to use a proper sell-signal - then you can optimize without trailing-stoploss, and use that to improve profits eventually.

**Comment by Fimap9to1** [comment]
Thanks!
